mergin/client_pull.py

```diff
@@ -551,3 +572,176 @@ def pull_project_finalize(job):
+    """
+    mc.log.info(f"--- start download {file_path} for {project_path}")
+    project_info = mc.project_info(project_path, version=version)
+
+    mc.log.info(f"Got project info. version {project_info['version']}")
+
+    # set temporary directory and make sure the destination directory exists
+    temp_dir = os.path.join(tempfile.gettempdir(), "mergin_temp")
+    os.makedirs(os.path.dirname(output_file), exist_ok=True)
+
+    download_list = []
+    update_tasks = []
+    total_size = 0
+    for file in project_info['files']:
+        if file["path"] == file_path:
+            file['version'] = version
+            items = _download_items(file, temp_dir)
+            task = UpdateTask(file['path'], items, output_file)
+            download_list.extend(task.download_queue_items)
+            for item in task.download_queue_items:
+                total_size += item.size
+            update_tasks.append(task)
+            break
+    if not download_list:
+        warn = f"No {file_path} exists at version {version}"
+        mc.log.warning(warn)
+        raise ClientError(warn)
+
+    mc.log.info(f"will download file {file_path} in {len(download_list)} chunks, total size {total_size}")
+    job = DownloadJob(
+        project_path, total_size, version, update_tasks, download_list, temp_dir, None, mc, project_info
+    )
+    job.executor = concurrent.futures.ThreadPoolExecutor(max_workers=4)
+    job.futures = []
+    for item in download_list:
+        future = job.executor.submit(_do_download, item, mc, None, project_path, job)
+        job.futures.append(future)
+
+    return job
+
+
+def download_file_finalize(job):
+    """
+    To be called when download_file_async is finished
+    """
+    job.executor.shutdown(wait=True)
+
+    # make sure any exceptions from threads are not lost
+    for future in job.futures:
+        if future.exception() is not None:
+            raise future.exception()
+
+    job.mc.log.info("--- download finished")
+
+    for task in job.update_tasks:
+        task.apply(job.directory, job.mp)
+
+
+def download_diffs_async(mc, project_directory, file_path, versions, file_history=None):
+    """
+    Starts background download project file diffs for specified versions.
+    Returns handle to the pending download.
+
+    Args:
+        mc (MerginClient): MerginClient instance.
+        project_directory (str): local project directory.
+        file_path (str): file path relative to Mergin project root.
+        versions ([str]): list of tags of versions to fetch, for example: ["v2", "v5"].
+        file_history (dict): optional file history info, result of MerginClient.project_file_history_info()
+
+    Returns:
+        PullJob/None: a handle for the pending download.
+    """
+    try:
+        mp = MerginProject(project_directory)
+    except InvalidProject as e:
+        mc.log.error(f"Couldn't create Mergin project for directory: {project_directory}: {repr(e)}")
+        return None
```

Comment:
it would be probably good enough to just leave InvalidProject to be raised rather than returning None (it is an error)